File: test/MilestonesTest.java
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class MilestonesTest {

  private static Event event(boolean external, boolean hideRegistration) {
    return new Event("e1", "Ein Talk (Marco Klaassen)", "Ein Talk", "abstract", "Marco Klaassen",
        "marco", "Darmstadt", "https://www.jug-da.de/talk/", "2026-08-20T18:30:00",
        external, hideRegistration);
  }

  private static Event event() {
    return event(false, false);
  }

  @Test
  @DisplayName("mail milestones cover the original exact day offsets")
  void coversOriginalOffsets() {
    assertEquals("announcement", Mail.due(event(), 28).orElseThrow().kind());
    assertEquals("invitation-7", Mail.due(event(), 7).orElseThrow().kind());
    assertEquals("invitation-2", Mail.due(event(), 2).orElseThrow().kind());
  }

  @Test
  @DisplayName("mail milestones catch up for a day or two after a missed run")
  void catchesUp() {
    assertEquals("announcement", Mail.due(event(), 26).orElseThrow().kind());
    assertEquals("invitation-7", Mail.due(event(), 5).orElseThrow().kind());
    assertEquals("invitation-2", Mail.due(event(), 1).orElseThrow().kind());
  }

  @Test
  @DisplayName("mail milestones never overlap or fire outside their window")
  void neverOverlaps() {
    for (var days : new int[] {40, 29, 25, 8, 4, 3, 0, -1}) {
      assertTrue(Mail.due(event(), days).isEmpty(), "unexpected mail at " + days + "d");
    }
  }

  @Test
  @DisplayName("external events never get a mail, but do get toots")
  void externalEventsAreTootOnly() {
    for (var days : new int[] {28, 7, 2}) {
      assertTrue(Mail.due(event(true, false), days).isEmpty());
    }
    assertEquals("toot-7", Toot.due(event(true, false), 7).orElseThrow().kind());
  }

  @Test
  @DisplayName("hideRegistration suppresses only the announcement")
  void hideRegistrationSuppressesAnnouncementOnly() {
    assertTrue(Mail.due(event(false, true), 28).isEmpty());
    assertEquals("invitation-7", Mail.due(event(false, true), 7).orElseThrow().kind());
  }

  @Test
  @DisplayName("toots also go out on the day itself, but only exactly then")
  void tootsOnTheDay() {
    assertEquals("toot-0", Toot.due(event(), 0).orElseThrow().kind());
    assertTrue(Toot.due(event(), -1).isEmpty());
  }

  @Test
  @DisplayName("the same-day toot says HEUTE, the others carry the date")
  void sameDayTootSaysHeute() {
    assertTrue(Toot.compose(event(), 0).startsWith("!!!HEUTE!!! "), Toot.compose(event(), 0));
    assertTrue(Toot.compose(event(), 7).startsWith("20.08.: Ein Talk"), Toot.compose(event(), 7));
  }

  @Test
  @DisplayName("the toot keeps the speaker and handle the template asks for")
  void tootCarriesSpeaker() {
    var toot = Toot.compose(event(), 7);
    assertTrue(toot.contains("(Marco Klaassen / @marco)"), toot);
    assertTrue(toot.contains("https://www.jug-da.de/talk/"), toot);
  }

  @Test
  @DisplayName("an inverted window is a programming error, caught at construction")
  void invertedWindowRejected() {
    assertThrows(IllegalArgumentException.class, () -> new Milestone("bad", "t", 10, 2));
  }
}

File: test/milestones.test.js
import assert from 'node:assert/strict';
import { test } from 'node:test';
import * as mail from '../mail.js';
import * as toot from '../toot.js';
import * as state from '../state.js';

const event = (extra = {}) => ({ uid: 'e1', start: '2026-08-20T18:30:00', summary: 's', ...extra });

test('mail milestones cover the old exact-day offsets', () => {
  assert.equal(mail.due(event(), 28).kind, 'announcement');
  assert.equal(mail.due(event(), 7).kind, 'invitation-7');
  assert.equal(mail.due(event(), 2).kind, 'invitation-2');
});

test('mail milestones catch up for a day or two after a missed run', () => {
  assert.equal(mail.due(event(), 26).kind, 'announcement');
  assert.equal(mail.due(event(), 5).kind, 'invitation-7');
  assert.equal(mail.due(event(), 1).kind, 'invitation-2');
});

test('mail milestones do not overlap or fire outside their window', () => {
  for (const diff of [40, 29, 25, 8, 4, 3, 0, -1]) {
    assert.equal(mail.due(event(), diff), null, `unexpected mail at ${diff}d`);
  }
});

test('external events never get a mail', () => {
  for (const diff of [28, 7, 2]) {
    assert.equal(mail.due(event({ externalEvent: true }), diff), null);
  }
});

test('hideRegistration suppresses only the announcement', () => {
  const ev = event({ hideRegistration: true });
  assert.equal(mail.due(ev, 28), null);
  assert.equal(mail.due(ev, 7).kind, 'invitation-7');
});

test('toots also go out on the day itself, but only exactly then', () => {
  assert.equal(toot.due(event(), 0).kind, 'toot-0');
  assert.equal(toot.due(event(), -1), null);
});

test('toots go to external events too', () => {
  assert.equal(toot.due(event({ externalEvent: true }), 7).kind, 'toot-7');
});

test('the same-day toot says HEUTE, the others carry the date', () => {
  assert.match(toot.compose(event(), 0), /^!!!HEUTE!!! /);
  assert.match(toot.compose(event({ title: 't' }), 7), /^20\.08\.: t/);
});

test('pruning keeps anything still in the feed', () => {
  const sent = new Map([['e1:toot-7', '2020-01-01']]);
  state.prune(sent, new Set(['e1']), new Date('2026-08-01'));
  assert.ok(sent.has('e1:toot-7'));
});

test('pruning keeps recent sends even once the event leaves the feed', () => {
  const sent = new Map([['gone:toot-7', '2026-07-20']]);
  state.prune(sent, new Set(), new Date('2026-08-01'));
  assert.ok(sent.has('gone:toot-7'));
});

test('pruning drops only old sends for events that are gone', () => {
  const sent = new Map([['gone:toot-7', '2026-01-01']]);
  state.prune(sent, new Set(), new Date('2026-08-01'));
  assert.equal(sent.size, 0);
});

test('an empty feed cannot wipe recent state', () => {
  const sent = new Map([['e1:toot-7', '2026-07-30'], ['e2:invitation-2', '2026-07-25']]);
  state.prune(sent, new Set(), new Date('2026-08-01'));
  assert.equal(sent.size, 2);
});
